ls1intum/Artemis
# `Development`: Track locked status for programming exercise participations

### Checklist
#### General
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
Logic for whether participations are locked due to the start date, the due date, or the submission limit, is scattered around the server and client code.

### Description
This PR adds a field `locked` to the  participation table that is kept up to date.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise without dates or submission limit configured, offline IDE should be allowed

1. Log in to Artemis as the student and participate in the exercise.
2. Log in as the instructor and edit the exercise, disallowing offline IDE usage.
3. Log in as the student again. Make sure you can still participate in the online editor but not by using your local Git client.
4. Log in as the instructor and edit the exercise again, e.g. by adding a due date in the past, or adding a submission limit (you cannot add the submission limit via the normal update exercise view, you need to go to the grading of the exercise for that)
5. Make sure the changes are reflected for the student. E.g. when you introduced a submission limit, make sure the student cannot submit after reaching that limit and "Repository is locked" is shown after reloading)

#### Exam Mode Testing
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Programming Exercise with a submission limit configured

1. Log in to Artemis
2. Participate in the exam as a student and submit until hitting the submission limit
3. Make sure you receive an error message that informs you about the submission limit. Reload the page and make sure the online editor is locked.
4. After handing in the exam, make sure you cannot push via your local Git client anymore.

### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage
#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exam-submission.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exercises/programming/programming-exam-submission.component.ts.html) | 88.88% | ok |
| [code-editor-actions.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/actions/code-editor-actions.component.ts.html) | 81.14% |ok |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ProgrammingExerciseStudentParticipation.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.participation/ProgrammingExerciseStudentParticipation.html) | 100% | ok |
| [ParticipationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ParticipationService.html) | 94% | ok |
| [RepositoryAccessService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/RepositoryAccessService.html) | 98% | ok |
| [SubmissionPolicyService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/SubmissionPolicyService.html) | 92% | ok |
| [ExamService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamService.html) | 95% | ok |
| [InstanceMessageReceiveService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/InstanceMessageReceiveService.html) | 75% | ok |
| [MainInstanceMessageSendService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/MainInstanceMessageSendService.html) | 96% | ok |
| [ProgrammingExerciseGradingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseGradingService.html) | 97% | ok |
| [ProgrammingExerciseParticipationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseParticipationService.html) | 70% | ok, all changes covered |
| [ProgrammingExerciseRepositoryService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseRepositoryService.html) | 92% | ok |
| [ProgrammingExerciseScheduleService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.scheduled/ProgrammingExerciseScheduleService.html) | 86% | ok |
| [ParticipationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ParticipationResource.html) | 91% | ok |
